### PR TITLE
Newsletter input component

### LINKED ISSUE #52: Newsletter input component
#### Wireframes
https://www.figma.com/file/yVO3Meu9PwRPhWI1JBtwKd/newsletter?node-id=0%3A1

-----

#### Used by
- All projects

#### Props
- mailchimpId

### PR DESCRIPTION
**Pull request by andreicioban**
Closes #52 

* Changed the `Input` component to support [addons](https://bulma.io/documentation/form/general/#form-addons)
* Added a `SubscribeForm` form
* Added `MailchimpSubscribe` component - wraps the subscribe form and posts to `mailchimp url`

To do:
- [x] Input validation
- [x] Error reporting
- [x] Success state/message
- [x] Support for compact layout

**To see it working you will need a valid Mailchimp URL**

subscribe-wide

Compact:

subs-compact

### CODE REVIEW

File: vim.exe.stackdump [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Stack trace:
```
Comment (dmfarcas) [inline vim.exe.stackdump]: Is this file needed? :D

Comment (andreicioban) [reply vim.exe.stackdump]: I have no idea where this came from.
Removed

**Comment by aniri** [comment]
@andreicioban can you please fix the merge conflicts?

File: src/components/subscribe/subscribe-form.jsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  success,
+  errorMessage,
+  onSubmitted
+}) => {
+  const [inputValue, setInputValue] = useState();
+  const [isValid, setIsValid] = useState();
+
+  const handleInputChange = ({ target }) => {
+    setIsValid(validateEmail(target.value));
+    setInputValue(target.value);
```
Comment (moonflare) [inline src/components/subscribe/subscribe-form.jsx]: nit
```suggestion
  const handleInputChange = ({ target: { value } }) => {
    setIsValid(validateEmail(value));
    setInputValue(value);
```

File: src/components/subscribe/mailchimp-subscribe.jsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      setIsSending(false);
+      setIsSuccess(true);
+    } catch (errorMsg) {
+      setErrorMessage(errorMsg);
+    }
+  };
+
+  return (
+    <SubscribeForm
+      onSubmitted={formData => subscribe(formData)}
```
Comment (moonflare) [inline src/components/subscribe/mailchimp-subscribe.jsx]: nit
```suggestion
      onSubmit={formData => subscribe(formData)}
```

File: src/components/subscribe/mailchimp-subscribe.stories.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+export const mailchimpSubscribe = () => {
+  const [chimpUrl, setChimpUrl] = useState();
+  return (
+    <Wrapper>
+      <label className="label">Mailchimp URL:</label>
+      <Input
+        type="text"
+        label="Introdu un url de mailchimp valid"
+        usePlaceholder={true}
+        onChange={({ target }) => setChimpUrl(target.value)}
```
Comment (moonflare) [inline src/components/subscribe/mailchimp-subscribe.stories.js]: nit
```suggestion
        onChange={({ target: { value } }) => setChimpUrl(value)}
```

File: src/components/subscribe/mailchimp-subscribe.stories.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  </div>
+);
+
+Wrapper.propTypes = {
+  children: PropTypes.node,
+  classes: PropTypes.string
+};
+
+export const mailchimpSubscribe = () => {
+  const [chimpUrl, setChimpUrl] = useState();
```
Comment (moonflare) [inline src/components/subscribe/mailchimp-subscribe.stories.js]: And we can remove the check from `L32`
```suggestion
  const [chimpUrl, setChimpUrl] = useState("");
```

File: src/components/subscribe/subscribe-form.jsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+SubscribeForm.propTypes = {
+  compact: PropTypes.bool,
+  sending: PropTypes.bool,
+  success: PropTypes.bool,
+  errorMessage: PropTypes.string,
+  onSubmitted: PropTypes.func
+};
+
+SubscribeForm.defaultProps = {
+  onSubmitted: ({ email }) => console.log(email),
```
Comment (moonflare) [inline src/components/subscribe/subscribe-form.jsx]: I guess we can remove this one
```suggestion
```

File: src/components/subscribe/mailchimp-subscribe.jsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  const subscribe = async ({ email }) => {
+    setIsSending(true);
+    try {
+      await sendSubscriptionRequest(url, email);
+      setIsSending(false);
+      setIsSuccess(true);
+    } catch (errorMsg) {
+      setErrorMessage(errorMsg);
+    }
```
Comment (moonflare) [inline src/components/subscribe/mailchimp-subscribe.jsx]: We should toggle `isSending` in a finally block so we're covered when the request fails as well.
```suggestion
    try {
      await sendSubscriptionRequest(url, email);
      setIsSuccess(true);
    } catch (errorMsg) {
      setErrorMessage(errorMsg);
    } finally {
      setIsSending(false); 
    }
```

Comment (moonflare) [reply src/components/subscribe/mailchimp-subscribe.jsx]: Should we also `setIsSuccess(false);` in the catch block?

File: src/components/subscribe/mailchimp-subscribe.jsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import React, { useState } from "react";
+import PropTypes from "prop-types";
+import { SubscribeForm } from "./subscribe-form";
+import { sendSubscriptionRequest } from "./mailchimp.service";
+
+export const MailchimpSubscribe = ({ url }) => {
+  const [isSending, setIsSending] = useState();
```
Comment (moonflare) [inline src/components/subscribe/mailchimp-subscribe.jsx]: nit
```suggestion
  const [isLoading, setIsLoading] = useState(false);
```

File: src/components/subscribe/subscribe-form.jsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  sending,
+  success,
+  errorMessage,
+  onSubmitted
+}) => {
+  const [inputValue, setInputValue] = useState();
+  const [isValid, setIsValid] = useState();
+
+  const handleInputChange = ({ target }) => {
+    setIsValid(validateEmail(target.value));
```
Comment (moonflare) [inline src/components/subscribe/subscribe-form.jsx]: Showing the validation while the user is still typing is not a very good pattern. It's called "Don't scream at the user". We should show the validation only after the user click on the CTA button 🙌

File: src/components/subscribe/mailchimp-subscribe.jsx (hunk 7 above)
Comment (andreicioban) [reply src/components/subscribe/mailchimp-subscribe.jsx]: success will become true only if the request succeeds. 
So no need IMHO.

File: src/components/subscribe/email-validator.service.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+const emailRegex = new RegExp("[^@]+@[^.]+..+");
```
Comment (moonflare) [inline src/components/subscribe/email-validator.service.js]: Do you know about this one, haha? https://emailregex.com/
